cranelift/assembler-x64/src/vex.rs

```diff
@@ -0,0 +1,259 @@
+//! Encoding logic for VEX instructions.
+use super::rex;
+use super::XmmMem;
+use crate::api::{AsReg, CodeSink, KnownOffsetTable, Registers};
+use crate::mem::emit_modrm_sib_disp;
+use crate::Amode;
+
+/// Allows using the same opcode byte in different "opcode maps" to allow for more instruction
+/// encodings. See appendix A in the Intel Software Developer's Manual, volume 2A, for more details.
+#[derive(PartialEq)]
+pub enum OpcodeMap {
+    None,
+    _0F,
+    _0F38,
+    _0F3A,
+}
+
+impl OpcodeMap {
+    /// Normally the opcode map is specified as bytes in the instruction, but some x64 encoding
+    /// formats pack this information as bits in a prefix (e.g. VEX / EVEX).
+    pub fn bits(&self) -> u8 {
+        match self {
+            OpcodeMap::None => 0b00,
+            OpcodeMap::_0F => 0b01,
+            OpcodeMap::_0F38 => 0b10,
+            OpcodeMap::_0F3A => 0b11,
+        }
+    }
+}
+
+impl Default for OpcodeMap {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+/// We may need to include one or more legacy prefix bytes before the REX prefix.  This enum
+/// covers only the small set of possibilities that we actually need.
+#[derive(PartialEq)]
+pub enum LegacyPrefix {
+    /// No prefix bytes.
+    None,
+    /// Operand Size Override -- here, denoting "16-bit operation".
+    _66,
+    /// The Lock prefix.
+    _F0,
+    /// Operand size override and Lock.
+    _66F0,
+    /// REPNE, but no specific meaning here -- is just an opcode extension.
+    _F2,
+    /// REP/REPE, but no specific meaning here -- is just an opcode extension.
+    _F3,
+    /// Operand size override and same effect as F3.
+    _66F3,
+}
+
+impl LegacyPrefix {
+    /// Emit the legacy prefix as bits (e.g. for EVEX instructions).
+    #[inline(always)]
+    pub(crate) fn bits(&self) -> u8 {
+        match self {
+            Self::None => 0b00,
+            Self::_66 => 0b01,
+            Self::_F3 => 0b10,
+            Self::_F2 => 0b11,
+            _ => panic!(
+                "VEX and EVEX bits can only be extracted from single prefixes: None, 66, F3, F2"
+            ),
+        }
+    }
+}
+
+impl Default for LegacyPrefix {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+pub struct VexInstruction<R: Registers> {
+    pub length: VexVectorLength,
+    pub prefix: LegacyPrefix,
+    pub map: OpcodeMap,
+    pub opcode: u8,
+    pub w: bool,
+    pub reg: u8,
+    pub rm: XmmMem<R::ReadXmm, R::ReadGpr>,
+    pub vvvv: Option<u8>,
+    pub imm: Option<u8>,
+}
+
+impl<R: Registers> Default for VexInstruction<R> {
+    fn default() -> Self {
+        Self {
+            length: VexVectorLength::default(),
+            prefix: LegacyPrefix::None,
+            map: OpcodeMap::None,
+            opcode: 0x00,
+            w: false,
+            reg: 0x00,
+            rm: XmmMem::default(),
+            vvvv: None,
+            imm: None,
+        }
+    }
+}
```

Comment:
We shouldn't need to create a default (i.e., invalid) VEX instruction ever, right?

Reply:
I don't think so. Will remove this.